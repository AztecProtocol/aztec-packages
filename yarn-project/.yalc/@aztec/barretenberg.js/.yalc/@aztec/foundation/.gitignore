dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs