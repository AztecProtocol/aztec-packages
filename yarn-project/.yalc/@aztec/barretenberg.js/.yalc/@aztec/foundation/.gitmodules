[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
