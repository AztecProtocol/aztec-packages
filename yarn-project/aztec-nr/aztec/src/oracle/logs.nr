use crate::types::point::Point;
use crate::constants_gen::NUM_FIELDS_PER_SHA256;

// TODO: Should take encrypted data.
#[oracle(emitEncryptedLog)]
fn emit_encrypted_log_oracle<N>(_contract_address: Field, _storage_slot: Field, _encryption_pub_key: Point, _preimage: [Field; N]) -> Field {}

unconstrained pub fn emit_encrypted_log<N>(contract_address: Field, storage_slot: Field, encryption_pub_key: Point, preimage: [Field; N]) -> [Field; NUM_FIELDS_PER_SHA256] {
    [emit_encrypted_log_oracle(contract_address, storage_slot, encryption_pub_key, preimage), 0]
}

#[oracle(emitUnencryptedLog)]
fn emit_unencrypted_log_oracle<T>(_contract_address: Field, _event_selector: Field, _message: T) -> Field {}

unconstrained pub fn emit_unencrypted_log<T>(contract_address: Field, event_selector: Field, message: T) -> [Field; NUM_FIELDS_PER_SHA256] {
    // https://github.com/AztecProtocol/aztec-packages/issues/885
    [emit_unencrypted_log_oracle(contract_address, event_selector, message), 0]
}
