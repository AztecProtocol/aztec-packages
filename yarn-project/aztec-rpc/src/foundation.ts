import nodeCrypto from 'crypto';
import { Keccak } from 'sha3';

export const randomBytes = (len: number) => {
  return nodeCrypto.randomBytes(len) as Buffer;
};

export function keccak256(input: Buffer | string) {
  const inputBuf = typeof input === 'string' ? Buffer.from(input) : input;
  const hash = new Keccak(256);
  return hash.update(inputBuf).digest();
}
