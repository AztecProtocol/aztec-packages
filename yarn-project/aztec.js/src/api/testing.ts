export { AnvilTestWatcher } from '../test/anvil_test_watcher.js';
export { RollupCheatCodes } from '../test/rollup_cheat_codes.js';
export { AztecCheatCodes } from '../test/aztec_cheat_codes.js';
export { EthCheatCodes } from '@aztec/ethereum/eth-cheatcodes';
export { CheatCodes } from '../test/cheat_codes.js';
