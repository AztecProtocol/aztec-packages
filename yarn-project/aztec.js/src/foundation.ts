export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export class Timer {
  private start: number;

  constructor() {
    this.start = new Date().getTime();
  }

  public ms() {
    return new Date().getTime() - this.start;
  }

  public s() {
    return (new Date().getTime() - this.start) / 1000;
  }
}

export function* backoffGenerator() {
  const v = [1, 1, 1, 2, 4, 8, 16, 32, 64];
  let i = 0;
  while (true) {
    yield v[Math.min(i++, v.length - 1)];
  }
}

export async function retry<Result>(fn: () => Promise<Result>, name = 'Operation', backoff = backoffGenerator()) {
  while (true) {
    try {
      return await fn();
    } catch (err: any) {
      const s = backoff.next().value;
      if (s === undefined) {
        throw err;
      }
      console.log(`${name} failed. Will retry in ${s}s...`);
      console.log(err);
      await sleep(s * 1000);
      continue;
    }
  }
}

// Call `fn` repeatedly until it returns true or timeout.
// Both `interval` and `timeout` are seconds.
// Will never timeout if the value is 0.
export async function retryUntil<T>(fn: () => Promise<T | undefined>, name = '', timeout = 0, interval = 1) {
  const timer = new Timer();
  while (true) {
    const result = await fn();
    if (result) {
      return result;
    }

    await sleep(interval * 1000);

    if (timeout && timer.s() > timeout) {
      throw new Error(name ? `Timeout awaiting ${name}` : 'Timeout');
    }
  }
}
