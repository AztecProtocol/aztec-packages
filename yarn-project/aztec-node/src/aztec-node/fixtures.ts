import {
  AccumulatedData,
  AffineElement,
  AggregationObject,
  ConstantData,
  ContractDeploymentData,
  EMITTED_EVENTS_LENGTH,
  EthAddress as CircuitEthAddress,
  Fq,
  Fr,
  FunctionData,
  KERNEL_L1_MSG_STACK_LENGTH,
  KERNEL_NEW_COMMITMENTS_LENGTH,
  KERNEL_NEW_CONTRACTS_LENGTH,
  KERNEL_NEW_NULLIFIERS_LENGTH,
  KERNEL_OPTIONALLY_REVEALED_DATA_LENGTH,
  KERNEL_PRIVATE_CALL_STACK_LENGTH,
  KERNEL_PUBLIC_CALL_STACK_LENGTH,
  NewContractData,
  OldTreeRoots,
  OptionallyRevealedData,
  PrivateKernelPublicInputs,
  TxContext,
  UInt8Vector,
} from '@aztec/circuits.js';
import { EthereumRpc } from '@aztec/ethereum.js/eth_rpc';
import { WalletProvider } from '@aztec/ethereum.js/provider';
import { AztecAddress, randomBytes } from '@aztec/foundation';
import { Rollup, Yeeter } from '@aztec/l1-contracts';
import { Tx } from '@aztec/tx';

// REFACTOR: Move deployment logic to l1-contracts package, and refactor it out of other integration tests (archiver, sequencer)
export const deployRollupContract = async (provider: WalletProvider, ethRpc: EthereumRpc) => {
  const deployAccount = provider.getAccount(0);
  const contract = new Rollup(ethRpc, undefined, { from: deployAccount, gas: 1000000 });
  await contract.deploy().send().getReceipt();
  return contract.address;
};

export const deployYeeterContract = async (provider: WalletProvider, ethRpc: EthereumRpc) => {
  const deployAccount = provider.getAccount(0);
  const contract = new Yeeter(ethRpc, undefined, { from: deployAccount, gas: 1000000 });
  await contract.deploy().send().getReceipt();
  return contract.address;
};

export const createProvider = (host: string, mnemonic: string, accounts: number) => {
  const walletProvider = WalletProvider.fromHost(host);
  walletProvider.addAccountsFromMnemonic(mnemonic, accounts);
  return walletProvider;
};

// REFACTOR: Use @aztec/circuit.js/factories where possible
export const createCircuitEthAddress = () => {
  return new CircuitEthAddress(randomBytes(20));
};

export const createRandomCommitments = (num: number) => {
  return Array(num)
    .fill(0)
    .map(() => new Fr(randomBytes(32)));
};

export const createOptionallyRetrievedData = () => {
  const func = new FunctionData(0, true, true);
  return new OptionallyRevealedData(
    new Fr(0),
    func,
    createRandomCommitments(EMITTED_EVENTS_LENGTH),
    new Fr(0),
    createCircuitEthAddress(),
    true,
    true,
    true,
    true,
  );
};

export const createOptionallyRetrievedDatas = (num: number) => {
  return Array(num)
    .fill(0)
    .map(() => createOptionallyRetrievedData());
};

export const createNewContractData = () => {
  return new NewContractData(AztecAddress.random(), createCircuitEthAddress(), new Fr(randomBytes(32)));
};

export const createNewContractDatas = (num: number) => {
  return Array(num)
    .fill(0)
    .map(() => createNewContractData());
};

export const createTx = () => {
  const oldTreeRoots = new OldTreeRoots(new Fr(0), new Fr(0), new Fr(0), new Fr(0));
  const contractDeploymentData = new ContractDeploymentData(
    new Fr(randomBytes(32)),
    new Fr(randomBytes(32)),
    new Fr(randomBytes(32)),
    createCircuitEthAddress(),
  );
  const txContext = new TxContext(false, false, true, contractDeploymentData);
  const constantData = new ConstantData(oldTreeRoots, txContext);
  const aggregationObject = new AggregationObject(
    new AffineElement(new Fq(0), new Fq(0)),
    new AffineElement(new Fq(0), new Fq(0)),
    [],
    [],
    false,
  );
  const accumulatedData = new AccumulatedData(
    aggregationObject,
    new Fr(0),
    createRandomCommitments(KERNEL_NEW_COMMITMENTS_LENGTH),
    createRandomCommitments(KERNEL_NEW_NULLIFIERS_LENGTH),
    createRandomCommitments(KERNEL_PRIVATE_CALL_STACK_LENGTH),
    createRandomCommitments(KERNEL_PUBLIC_CALL_STACK_LENGTH),
    createRandomCommitments(KERNEL_L1_MSG_STACK_LENGTH),
    createNewContractDatas(KERNEL_NEW_CONTRACTS_LENGTH),
    createOptionallyRetrievedDatas(KERNEL_OPTIONALLY_REVEALED_DATA_LENGTH),
  );
  const kernelInputs = new PrivateKernelPublicInputs(accumulatedData, constantData, true);
  return new Tx(kernelInputs, new UInt8Vector(Buffer.alloc(0)));
};
