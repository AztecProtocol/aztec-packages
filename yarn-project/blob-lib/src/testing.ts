import { BLOBS_PER_BLOCK } from '@aztec/constants';
import { makeTuple } from '@aztec/foundation/array';
import { toBufferBE } from '@aztec/foundation/bigint-buffer';
import { Fr } from '@aztec/foundation/fields';

import { Blob } from './blob.js';
import { BlobPublicInputs, BlockBlobPublicInputs } from './blob_public_inputs.js';
import { TX_START_PREFIX, TX_START_PREFIX_BYTES_LENGTH } from './encoding.js';
import { Poseidon2Sponge, SpongeBlob } from './sponge_blob.js';

/**
 * Makes arbitrary poseidon sponge for blob inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the sponge.
 * @returns A sponge blob instance.
 */
export function makeSpongeBlob(seed = 1): SpongeBlob {
  return new SpongeBlob(
    new Poseidon2Sponge(
      makeTuple(3, i => new Fr(i)),
      makeTuple(4, i => new Fr(i)),
      1,
      false,
    ),
    seed,
    seed + 1,
  );
}

/**
 * Makes arbitrary blob public inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the blob inputs.
 * @returns A blob public inputs instance.
 */
export function makeBlobPublicInputs(seed = 1): BlobPublicInputs {
  return new BlobPublicInputs(
    new Fr(seed),
    BigInt(seed + 1),
    makeTuple(2, i => new Fr(i)),
  );
}

/**
 * Makes arbitrary block blob public inputs.
 * Note: will not verify inside the circuit.
 * @param seed - The seed to use for generating the blob inputs.
 * @returns A block blob public inputs instance.
 */
export function makeBlockBlobPublicInputs(seed = 1): BlockBlobPublicInputs {
  return new BlockBlobPublicInputs(makeTuple(BLOBS_PER_BLOCK, () => makeBlobPublicInputs(seed)));
}

// TODO: copied form stdlib tx effect
function encodeFirstField(length: number): Fr {
  const lengthBuf = Buffer.alloc(2);
  lengthBuf.writeUInt16BE(length, 0);
  return new Fr(
    Buffer.concat([
      toBufferBE(TX_START_PREFIX, TX_START_PREFIX_BYTES_LENGTH),
      Buffer.alloc(1),
      lengthBuf,
      Buffer.alloc(1),
      Buffer.from([1]),
      Buffer.alloc(1),
      Buffer.alloc(1),
    ]),
  );
}

/**
 * Make an encoded blob with the given length
 *
 * This will deserialise correctly in the archiver
 * @param length
 * @returns
 */
export function makeEncodedBlob(length: number): Promise<Blob> {
  return Blob.fromFields([encodeFirstField(length + 1), ...Array.from({ length: length }, () => Fr.random())]);
}

/**
 * Make an unencoded blob with the given length
 *
 * This will fail deserialisation in the archiver
 * @param length
 * @returns
 */
export function makeUnencodedBlob(length: number): Promise<Blob> {
  return Blob.fromFields([...Array.from({ length: length }, () => Fr.random())]);
}

export function makeEncodedBlobFields(fields: Fr[]): Promise<Blob> {
  return Blob.fromFields([encodeFirstField(fields.length + 1), ...fields]);
}
