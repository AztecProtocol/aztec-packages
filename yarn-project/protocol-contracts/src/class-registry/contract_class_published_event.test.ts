import { setupCustomSnapshotSerializers } from '@aztec/foundation/testing';
import { ContractClassLog } from '@aztec/stdlib/logs';

import { getSampleContractClassPublishedEventPayload } from '../tests/fixtures.js';
import { ContractClassPublishedEvent } from './contract_class_published_event.js';

describe('ContractClassPublishedEvent', () => {
  beforeAll(() => setupCustomSnapshotSerializers(expect));
  it('parses an event as emitted by the ContractClassRegistry', () => {
    const log = ContractClassLog.fromBuffer(getSampleContractClassPublishedEventPayload());
    expect(ContractClassPublishedEvent.isContractClassPublishedEvent(log)).toBe(true);

    const event = ContractClassPublishedEvent.fromLog(log);

    // See ./__snapshots__/README.md for how to update the snapshot.
    expect(event).toMatchSnapshot();
  });
});
