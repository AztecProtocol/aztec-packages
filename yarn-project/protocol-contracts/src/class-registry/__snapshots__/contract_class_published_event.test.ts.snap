// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ContractClassPublishedEvent parses an event as emitted by the ContractClassRegistry 1`] = `
ContractClassPublishedEvent {
  "artifactHash": Fr<0x2de8d9b832a20a270f42ce01cf1bb2ab4ee28b6aaa9cef633a221c76e158219c>,
  "contractClassId": Fr<0x0a8ad282f9ed12cd5f61b8b0224b526b5950ebac30e6660189cb34418b6a3c66>,
  "packedPublicBytecode": Buffer<0x27000204012800000104804e2700000403270202040127020304001f0a00020003804d2e08804d00012500000045250000008e2802000104804e27020204003b0e000200012900804304ffffffff2800804404000328008045010000280080460400002800804700000028008048010001280080490400012800804a0000012800804b0400022800804c040004262500000d7929020002001e0149bf0a2a01020327020404002702060403002a0406052d0801020008010501270302040100220202052d0e040500220502052d0e04052702050403002a0205042902000400d9b5157824020003000000f023000006972d08010527020604040008010601270305040100220502061f308044804900062d08010600000102012d0e05062d08010500000102012e0c8046000527020804092d0800092d0a060a2d0a050b00080008002500000da22d0200002d0a0a0701220007804900092d0b0908270209040a2d08000a2d0a060b2d0a050c00080009002500000da22d0200002d0a0b0727020a040b2d08000b2d0a060c2d0a050d0008000a002500000da22d0200002d0a0c0901220009804900062d0b06051e020006002d08010927020a04030008010a012703090401002209020a360e0006000a00012200098049000b2d0b0b0a01220009804b000c2d0b0c0b1c0a0a0900042a090b0c2402000a0000020427020904003c0609012d08010927020a04030008010a012703090401002209020a360e0006000a02012200098049000a2d0b0a0601220009804b000b2d0b0b0a1c0a060900042a090a0b240200060000025627020904003c0609012d08010627020904020008010901270306040100220602091f30804980460009012200068049000a2d0b0a091c0a090a041c0a0a06002d080109000001020127020a004c2d0e0a092d08010a27020d04040008010d0127030a040100220a020d1f3080448049000d2d0b0a0d00220d020d2d0e0d0a2d08010d000001020127020e002c2d08010f2702100405000801100127030f040100220f02102d0a10112d0e0e1100221102112e0c8047001100221102112e0c8047001100221102112e0c804700112d0e0f0d2e088046000323000003290d220003804400072402000700000d17230000033e2d0b0d072b02000a000000000000000004000000000000000027021104122d0800122d0a0a1300080011002500000e192d0200002d0a130d2d0a140e2d0a150f2d0a16102d08010a00000102012d0e0d0a2d08010d00000102012d0e0e0d2d08010e00000102012d0e0f0e2d08010f00000102012d0e100f2d0b071000221002102d0e10072e088046000323000003ce0d220003804c00102402001000000ccd23000003e327021004112d0800112d0a0a122d0a0d132d0a0e142d0a0f1500080010002500000eac2d0200002d0a12072d0b090a2d0e0a09270209000d2d08010a27020d04040008010d0127030a040100220a020d2d0a0d0e2d0e090e00220e020e2d0e060e00220e020e2d0e070e2b020006000000000000000003000000000000000027020f04102d0800102d0a06110008000f002500000e192d0200002d0a11072d0a12092d0a130d2d0a140e2d08010600000102012d0e07062d08010700000102012d0e09072d08010900000102012d0e0d092d08010d00000102012d0e0e0d2d0b0a0e00220e020e2d0e0e0a2e088046000323000004d90d2200038044000e2402000e00000c7123000004ee27020a040e2d08000e2d0a060f2d0a07102d0a09112d0a0d120008000a002500000eac2d0200002d0a0f030a2a0b0306240200060000052b2500000f200b22000c804700031e020006010a2a0c0607122a030706240200060000054f2500000f321e02000300270207040327020a0403002a070a092d0801060008010901270306040100220602092d0e070900220902092d0e07092702090403002a0609072d0a07092d0e040900220902092d0e080900220902092d0e050900220602082d0b08072702090402002a0809053903a08043804300030007000520020003210200052d080107002207020a2d0b0a0927020b0402002a0a0b0822320005804600082d0a050927020b0403002a090b0a0008010a012703070401002207020b2d0e090b00220b020b2d0e090b2d0a090606220602062402000300000657230000062e2d0b070300220302032d0e030700220702082d0b08052702090402002a0809033c0e050323000006570b22000680460003240200030000067027020504003c0605011e020003003402000300220202062d0b06052702070402002a0607033b0e00050003230000069729020003005bd9f2da0a2a0103052d0b020300220302032d0e0302270203000224020005000006c4230000082a2d08010527020604030008010601270305040100220502061f30804b804900062d08010600000102012d0e05062d08010500000102012e0c8046000527020804092d0800092d0a060a2d0a050b00080008002500000f442d0200002d0a0a0701220007804900092d0b0908270209040a2d08000a2d0a060b2d0a050c00080009002500000f442d0200002d0a0b0701220007804900062d0b06051e020006001e02000700332a000600070009240200090000077d2500000fbb2d08010627020704030008010701270306040100220602072d0a07092d0e030900220902092d0e080927020804092d0800092d0a060a2e08804b000b2e088045000c00080008002500000fcd2d0200002d0a0a070b220007804700060b2200068045000824020008000007ee25000011492f0a00070006002a060508300a000800072d0b020500220502052d0e050200220202072d0b07062702080402002a0708053b0e00060005230000082a0a2a010405240200050000083c23000009762d08010427020504030008010501270304040100220402051f30804b804900052d08010500000102012d0e04052d08010400000102012e0c8046000427020704082d0800082d0a05092d0a040a00080007002500000f442d0200002d0a090601220006804900082d0b080727020804092d0800092d0a050a2d0a040b00080008002500000f442d0200002d0a0a0601220006804900052d0b05042d08010527020604030008010601270305040100220502062d0a06082d0e030800220802082d0e070827020704082d0800082d0a05092e08804b000a2e088045000b00080007002500000fcd2d0200002d0a09060b220006804700050b22000580450007240200070000094725000011492f0a00060005002a050407300a0007000600220202062d0b06052702070402002a0607043b0e0005000423000009762902000200bb19097e0a2a01020424020004000009912300000aa52d08010227020404020008010401270302040100220202041f3080498049000401220002804900052d0b05041e0200020924020002000009cf250000115b2d08010227020504030008010501270302040100220202052d0a05062d0e030600220602062d0e040627020404052d0800052d0a02062e08804b00072e088045000800080004002500000fcd2d0200002d0a06030b220003804700020b220002804500042402000400000a4025000011492f0a0003000227020404012702060403002a0406052d0801030008010501270303040100220302052d0e040500220502052d0e04052702050403002a0305042d0a04052d0e020500220302052d0b05042702060402002a0506023b0e000400022300000aa52702020255270203026e270204026b270205026f270206027727020702202702080273270209026527020a026c27020b026327020c027427020d027227020e027b27020f027d2d080110270211041c0008011101270310040100221002112d0a11122d0e021200221202122d0e031200221202122d0e041200221202122d0e031200221202122d0e051200221202122d0e061200221202122d0e031200221202122d0e071200221202122d0e081200221202122d0e091200221202122d0e0a1200221202122d0e091200221202122d0e0b1200221202122d0e0c1200221202122d0e051200221202122d0e0d1200221202122d0e071200221202122d0e0e1200221202122d0e081200221202122d0e091200221202122d0e0a1200221202122d0e091200221202122d0e0b1200221202122d0e0c1200221202122d0e051200221202122d0e0d1200221202122d0e0f120b208045804800022402000200000c71270203041e2d080104270205041e00080105012d0a04052a03000505274648b2f54117bd00220502050022100206270207041b2e02000680032e02000580042e0200078005250000116d270206041b002a0506052e0c804a000500220502052d0e010500220502053c0e03042402000e00000c7e2300000cbc00220a020f002a0f03102d0b100e27020f04102d0800102d0a06112d0a07122d0a09132d0a0d142d0a0e150008000f0025000011b32d0200002300000cbc012200038049000e2d0a0e0323000004d90022070211002a1103122d0b121027021104122d0800122d0a0a132d0a0d142d0a0e152d0a0f162d0a1017000800110025000011b32d02000001220003804900102d0a100323000003ce012200038049000700220a020f002a0f03102d0b100e2d0b0d0f0d220007804c00102402001000000d4625000012d72e02000f80032800800404000525000012e92e08800500100022100211002a1107122d0e0e122d0e100d2d0a07032300000329280080040478000d000000800480032400800300000da12a01000105f7a1f3afa5add4ca3c040201262500000d792d0b01032d0b02040d220004804400052402000500000dc425000012d70022030206002a0604072d0b07052d08010627020704020008010701270306040100220602072d0a07082d0e050801220004804900050e2a0405072402000700000e0c25000013772d0e03012d0e05022d0a0601262500000d792d08010227020304040008010301270302040100220202032d0a03042e0c8047000400220402042e0c8047000400220402042e0c804700042d08010327020404050008010401270303040100220302042d0a04052e0c8047000500220502052e0c8047000500220502052e0c8047000500220502052d0e01052e08804500042d0a02012d0a03022e0880460003262500000d792d0b04050b220005804500062402000600000ece27020704003c06070127020504062d0800062d0a01072d0a02082d0a03092d0a040a000800050025000013892d0200002d0b01052d0b02062d0b03072d0e05012d0e06022d0e07032e0c8048000401220006804900022d0b0201262a01000105f48001a659d327423c040201262a010001051f005012402422ee3c040201262500000d792d0b01032d0b02040d220004804b00052402000500000f6625000012d70022030206002a0604072d0b07052d08010627020704020008010701270306040100220602072d0a07082d0e050801220004804900050e2a0405072402000700000fae25000013772d0e03012d0e05022d0a0601262a01000105be1e3fff3ea4f6fa3c040201262500000d791c0a0205002b0200060000000000000000010000000000000000042a05060727020a040b2d08000b2d0a070c0008000a002500000e192d0200002d0a0c052d0a0d062d0a0e082d0a0f092d08010700000102012d0e05072d08010500000102012d0e06052d08010600000102012d0e08062d08010800000102012d0e09082d0b010900220902092d0e09012e088046000423000010680d220004804b000924020009000010e8230000107d240200030000108a23000010bc27020104092d0800092d0a070a2d0a050b2d0a060c2d0a080d2e08804a000e000800010025000011b32d02000023000010bc27020204092d0800092d0a070a2d0a050b2d0a060c2d0a080d00080002002500000eac2d0200002d0a0a01260c2a04020924020009000010fa2300001138002201020a002a0a040b2d0b0b0927020a040b2d08000b2d0a070c2d0a050d2d0a060e2d0a080f2d0a09100008000a0025000011b32d020000230000113801220004804900092d0a090423000010682a0100010502dc6e278076129d3c040201262a01000105fe3c8c0aa82dabf13c0402012601008003800580072e00800380082e00800480090b0080088007800a2400800a000011b22e01800880062e0480068009010080080002800801008009000280092300001181262500000d792d0b04060b2200068045000724020007000011d527020804003c0608012d0b03060b22000680440007240200070000126823000011ee2d0b03062d0b01072d0b02082d0b04090d2200068044000a2402000a0000121325000012d72e02000780032800800404000425000012e92e088005000a00220a020b002a0b060c2d0e050c01220006804900050e2a060507240200070000125325000013772d0e0a012d0e08022d0e05032d0e090423000012d627020604072d0800072d0a01082d0a02092d0a030a2d0a040b000800060025000013892d0200002d0b01062d0b02072d0b04082e02000680032800800404000425000012e92e0880050009012200098049000a2d0e050a2d0e09012d0e07022e0c804900032d0e080423000012d6262a01000105c56bc45a0e1000023c040201262e01800380060b008006000280072400800700001304230000130f2e008003800523000013762e0000018005010000018004000101008003800480092e008003800a2e008005800b0b00800a8009800c2400800c000013622e01800a80082e048008800b0100800a0002800a0100800b0002800b23000013312801800504000103008006000280062300001376262a0100010545a7ca711941e4153c040201262500000d792e088046000523000013990d22000580440006240200060000140423000013ae2d0b02052d0b050600220602062d0e06052d0801062702070405000801070127030604010022050207270208040400220602093f0f000700092d0b01052d0b03072d0b04082d0e05012d0e06022d0e07032d0e0804262d0b03060c2a050607240200070000141a23000014862d0b02060022060208002a0805092d0b09072d0b0108002208020a002a0a050b2d0b0b09002a07090a2d0b03072d0b04092e02000680032800800404000525000012e92e088005000b00220b020c002a0c050d2d0e0a0d2d0e08012d0e0b022d0e07032d0e0904230000148601220005804900062d0a06052300001399>,
  "privateFunctionsRoot": Fr<0x1922e7c1bb3c478e47f9dccff067dd9936c20492f82766ed61485882a8c32cc1>,
  "version": 1,
}
`;
